### DISCUSSION TITLE
0.5->0.6 migration advice?

### DISCUSSION THREAD

**Discussion by mock1328** [gtk-rs/gtk4-rs#1799] [Q&A] [answered]
For study purpose, I tried upgrading the package that I use, regreet, onto newer `gtk-rs` version. However, everything above `0.5` fails to compile, [this](https://github.com/rharish101/ReGreet/blob/1b3b53278d9df5b9d48d7a9f467e199856b8c313/src/gui/component.rs#L137) `view!` components, to be specific. It can't find the `template_child` elements, even though vscode manages to find them correctly into the `templates.rs` file, and these have correct `#[name]` attribute set up:

(example of error for `user_toggle`
```
cannot find value `user_toggle` in this scope
not found in this scope
```

If I try to upgrade to `0.7` or higher versions, some of compilation errors go away, but some related to `connect_clicked` are not:

e.g. for this code:

```rust
                    connect_activate[
                        sender, usernames_box, username_entry, sessions_box, session_entry
                    ] => move |this| {
                        sender.input(Self::Input::Login {
                            input: this.text().to_string(),
                            info: UserSessInfo::extract(
                                &usernames_box, &username_entry, &sessions_box, &session_entry
                            ),
                        })
                    }
```

I have an error:

```
cannot find value `usernames_box` in this scope
component.rs(229, 34): the binding `usernames_box` is available in a different scope in the same function
```

Thourgh trial-and-error I've managed to fix (?) it by making these references with `&`:

```rust
                    connect_activate[
                        sender, &usernames_box, &username_entry, &sessions_box, &session_entry
                    ] => move |this| {
```

However, now it errors on `move` statement:

```
used binding `usernames_box` isn't initialized
`usernames_box` used here but it isn't initialized
...
```

I've read the [blog](https://gtk-rs.org/blog/) entries for each release after `0.5`, read [changelog](https://github.com/gtk-rs/gtk4-rs/blob/master/CHANGELOG.md) and [release notes](https://github.com/gtk-rs/gtk4-rs/releases/tag/0.6.6), scrolled through the [comparison between tags](https://github.com/gtk-rs/gtk4-rs/compare/0.5.6...0.6.0) to see the changes in examples, but it still absolutely not clear to me why is it failing to compile and how can I fix it...

Not asking for a rewrite for me ofc, a simple advice of what's going on wrong would be enough :)

**Comment by sdroege** [comment, marked as answer]
This is probably something to ask somewhere related to relm4.

**Reply by mock1328** [reply]
Ah, I guess you're right, at least I can already see the fix for what's been fixed for me when I changed version to `0.7.0` instead of `0.6.0`: https://github.com/Relm4/Relm4/blob/main/CHANGES.md#062---2023-8-27
Hope note for latter is there too
